cmd/populator-controller/populator-controller.go

```diff
@@ -0,0 +1,97 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"github.com/konveyor/forklift-controller/pkg/apis/forklift/v1beta1"
+	populator_machinery "github.com/kubev2v/lib-volume-populator/populator-machinery"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+)
+
+const (
+	prefix     = "forklift.konveyor.io"
+	mountPath  = "/mnt/"
+	devicePath = "/dev/block"
+)
+
+const (
+	groupName  = "forklift.konveyor.io"
+	apiVersion = "v1beta1"
+	kind       = "OvirtVolumePopulator"
+	resource   = "ovirtvolumepopulators"
+	configName = "forklift-populator-controller"
+)
+
+func main() {
+	var httpEndpoint, metricsPath, masterURL, kubeconfig, imageName, namespace string
+
+	// Controller args
+	if f := flag.Lookup("kubeconfig"); f != nil {
+		kubeconfig = f.Value.String()
+	} else {
+		flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+	}
+	flag.StringVar(&masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	flag.StringVar(&imageName, "image-name", "", "Image to use for populating")
+	// Metrics args
+	flag.StringVar(&httpEndpoint, "http-endpoint", "", "The TCP network address where the HTTP server for diagnostics, including metrics and leader election health check, will listen (example: `:8080`). The default is empty string, which means the server is disabled.")
+	flag.StringVar(&metricsPath, "metrics-path", "/metrics", "The HTTP path where prometheus metrics will be exposed. Default is `/metrics`.")
+	// Other args
+	flag.StringVar(&namespace, "namespace", "konveyor-forklift", "Namespace to deploy controller")
+	flag.Parse()
+
+	var (
+		gk  = schema.GroupKind{Group: groupName, Kind: kind}
+		gvr = schema.GroupVersionResource{Group: groupName, Version: apiVersion, Resource: resource}
+	)
+	imageName = getImageName(kind, namespace)
+	populator_machinery.RunController(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath,
+		namespace, prefix, gk, gvr, mountPath, devicePath, getPopulatorPodArgs)
+}
+
+func getImageName(kind, namespace string) string {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		klog.Fatal(err.Error())
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		klog.Fatal(err.Error())
+	}
+
+	configMap, err := clientset.CoreV1().ConfigMaps(namespace).Get(context.TODO(), configName, metav1.GetOptions{})
```

Comment:
who is expected to create this configmap and what's the benefit compared to specifying it on the CR?

Reply:
specifying in the CR is tricky, because the CR is available only during the migration and the volpop controller is always running so there will be nothing to read most of the time

Reply:
we should deploy the configMap.

Reply:
> we should deploy the configMap.

yeah, I understand it's forklift - but do you expect the operator to do that?

> specifying in the CR is tricky, because the CR is available only during the migration and the volpop controller is always running so there will be nothing to read most of the time

I don't understand that part, I envisioned it as:
1. forklift-controller posts the CR, including the image that it gets from its env that was populated from ForkliftController
2. populator-controller reads the CR, including the image
3. populator-controller starts ovirt/osp-populator with the right image

do I understand this correctly that the problem with this is that it requires changing the library that k8s post as an example/utility for using volume-populator?

Reply:
> I don't understand that part, I envisioned it as:
> 
> 1. forklift-controller posts the CR, including the image that it gets from its env that was populated from ForkliftController
> 2. populator-controller reads the CR, including the image
> 3. populator-controller starts ovirt/osp-populator with the right image
> 
> do I understand this correctly that the problem with this is that it requires changing the library that k8s post as an example/utility for using volume-populator?

yes, the controller starts watching for PVCs with a specific `dataSourceRef` that matches the `gk`+`gvr` values we pass to `RunController`, so it needs know which CRs it's watching and which image to use when starting, but we can add this logic in our code and start multiple controllers, one for each kind with the correct image. We can probably optimize it in the future and only start a certain controller when the relevant provider is added

Reply:
>well yeah but this just exchanges listening to the populator CRs with listening to providers, no?
yeah... very optional in my opinion

Reply:
configmap idea dropped.